File: py-fake-service/app/adapter/mail_service.py
#
# Copyright (c) 2014 ThoughtWorks, Inc.
#
# Pixelated is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Pixelated is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with Pixelated. If not, see <http://www.gnu.org/licenses/>.
import os
import mailbox

from tagsset import TagsSet
from mailset import MailSet
from contacts import Contacts
from mail import Mail


class MailService:
    MAILSET_PATH = os.path.join(os.environ['HOME'], 'mailsets', 'mediumtagged')

    def __init__(self):
        self.mailset = MailSet()
        self.tagsset = TagsSet()
        self.contacts = Contacts()

    def reset(self):
        self.mailset = MailSet()
        self.tagsset = TagsSet()
        self.contacts = Contacts()

    def _read_file(self, filename):
        with open(filename, 'r') as fd:
            return fd.read()

    def _create_message_from_file(self, filename):
        data = self._read_file(filename)
        if data.startswith('From '):
            msg = mailbox.mbox(filename).popitem()[1]
            msg.from_addr = msg.get_from()
        else:
            msg = mailbox.Message(data)
            msg.from_addr = msg.get('From')
        return msg

    def load_mailset(self):
        mbox_filenames = [
            filename
            for filename in os.listdir
            (self.MAILSET_PATH) if filename.startswith('mbox')]
        messages = (self._create_message_from_file(os.path.join(self.MAILSET_PATH, mbox))
                 for mbox in mbox_filenames)

        for message in messages:
            self.mailset.add(message)
            self.tagsset.add(message)
            self.contacts.add(message)

    def mails(self, query, page, window_size):
        mails = self.mailset.values()
        mails = [mail for mail in mails if query.test(mail)]
        return sorted(mails, key=lambda mail: mail.date, reverse=True)

    def mail(self, mail_id):
        return self.mailset.get(mail_id)

    def search_contacts(self, query):
        return self.contacts.search(query)

    def mark_as_read(self, mail_id):
        self.mailset.mark_as_read(mail_id)
        self.tagsset.mark_as_read(self.mail(mail_id).tags)

    def delete_mail(self, mail_id):
        purged = self.mailset.delete(mail_id)
        if not purged:
            self.tagsset.increment_tag_total_count('trash')

    def update_tags_for(self, mail_id, new_tags):
        mail = self.mail(mail_id)

        new_tags_set = set(new_tags)
        old_tags_set = set(mail.tags)

        increment_set = new_tags_set - old_tags_set
        decrement_set = old_tags_set - new_tags_set

        map(lambda x: self.tagsset.increment_tag_total_count(x), increment_set)
        map(lambda x: self.tagsset.decrement_tag_total_count(x), decrement_set)

        mail.tags = new_tags

    def send(self, mail):
        mail = Mail.from_json(mail)
        self.mailset.update(mail)
        self.tagsset.increment_tag_total_count('sent')
        self.tagsset.decrement_tag_total_count('drafts')
        return mail.ident

    def save_draft(self, mail):
        mail = self.mailset.add_draft(Mail.from_json(mail))
        return mail.ident

    def update_draft(self, mail):
        mail = Mail.from_json(mail)
        self.mailset.update(mail)
        return mail.ident

    def draft_reply_for(self, mail_id):
        return self.mailset.find(draft_reply_for=mail_id)

File: web-ui/app/js/services/mail_service.js
/*
 * Copyright (c) 2014 ThoughtWorks, Inc.
 *
 * Pixelated is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Pixelated is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Pixelated. If not, see <http://www.gnu.org/licenses/>.
 */

define(
  [
    'flight/lib/component',
    'views/i18n',
    'services/model/mail',
    'helpers/monitored_ajax',
    'page/events',
    'features',
    'mixins/with_auto_refresh',
    'page/router/url_params'
  ], function (defineComponent, i18n, Mail, monitoredAjax, events, features, withAutoRefresh, urlParams) {

    'use strict';

    return defineComponent(mailService, withAutoRefresh('refreshMails'));

    function mailService() {
      var that;

      this.defaultAttrs({
        mailsResource: '/mails',
        singleMailResource: '/mail',
        currentTag: '',
        lastQuery: '',
        currentPage: 1,
        numPages: 1,
        pageSize: 25
      });

      this.errorMessage = function (msg) {
        return function () {
          that.trigger(document, events.ui.userAlerts.displayMessage, { message: msg });
        };
      };

      this.updateTags = function (ev, data) {
        var ident = data.ident;

        var success = function (data) {
          this.refreshMails();
          $(document).trigger(events.mail.tags.updated, { ident: ident, tags: data.tags });
          $(document).trigger(events.dispatchers.tags.refreshTagList, { skipMailListRefresh: true });
        };

        var failure = function (resp) {
          var msg = i18n('Could not update mail tags');
          if (resp.status === 403) {
            msg = i18n('Invalid tag name');
          }
          this.trigger(document, events.ui.userAlerts.displayMessage, { message: msg });
        };

        monitoredAjax(this, '/mail/' + ident + '/tags', {
          type: 'POST',
          contentType: 'application/json; charset=utf-8',
          data: JSON.stringify({newtags: data.tags})
        }).done(success.bind(this)).fail(failure.bind(this));

      };

      this.readMail = function (ev, data) {
        var mailIdents;
        if (data.checkedMails) {
          mailIdents = _.map(data.checkedMails, function (mail) {
            return mail.ident;
          });
        } else {
          mailIdents = [data.ident];
        }
        monitoredAjax(this, '/mails/read', {
          type: 'POST',
          data: JSON.stringify({idents: mailIdents})
        }).done(this.triggerMailsRead(data.checkedMails));
      };

      this.unreadMail = function (ev, data) {
        var mailIdents;
        if (data.checkedMails) {
          mailIdents = _.map(data.checkedMails, function (mail) {
            return mail.ident;
          });
        } else {
          mailIdents = [data.ident];
        }
        monitoredAjax(this, '/mails/unread', {
          type: 'POST',
          data: JSON.stringify({idents: mailIdents})
        }).done(this.triggerMailsRead(data.checkedMails));
      };

      this.triggerMailsRead = function (mails) {
        return _.bind(function () {
          this.refreshMails();
          this.trigger(document, events.ui.mails.uncheckAll);
        }, this);
      };

      this.triggerDeleted = function (dataToDelete) {
        return _.bind(function () {
          var mails = dataToDelete.mails || [dataToDelete.mail];

          this.refreshMails();
          this.trigger(document, events.ui.userAlerts.displayMessage, { message: dataToDelete.successMessage});
          this.trigger(document, events.ui.mails.uncheckAll);
          this.trigger(document, events.mail.deleted, { mails: mails });
        }, this);
      };

      this.deleteMail = function (ev, data) {
        monitoredAjax(this, '/mail/' + data.mail.ident,
          {type: 'DELETE'})
          .done(this.triggerDeleted(data))
          .fail(this.errorMessage(i18n('Could not delete email')));
      };

      this.deleteManyMails = function (ev, data) {
        var dataToDelete = data;
        var mailIdents = _.map(data.mails, function (mail) {
          return mail.ident;
        });

        monitoredAjax(this, '/mails/delete', {
          type: 'POST',
          dataType: 'json',
          contentType: 'application/json; charset=utf-8',
          data: JSON.stringify({idents: mailIdents})
        }).done(this.triggerDeleted(dataToDelete))
          .fail(this.errorMessage(i18n('Could not delete emails')));
      };

      function compileQuery(data) {
        var query = 'tag:"' + that.attr.currentTag + '"';

        if (data.tag === 'all') {
          query = 'in:all';
        }
        return query;
      }

      this.fetchByTag = function (ev, data) {
        this.attr.currentTag = data.tag;
        this.attr.lastQuery = compileQuery(data);
        this.updateCurrentPageNumber(1);

        this.refreshMails();
      };

      this.newSearch = function (ev, data) {
        this.attr.lastQuery = data.query;
        this.attr.currentTag = 'all';
        this.refreshMails();
      };

      this.mailFromJSON = function (mail) {
        return Mail.create(mail);
      };

      this.parseMails = function (data) {
        data.mails = _.map(data.mails, this.mailFromJSON, this);

        return data;
      };

      function escaped(s) {
        return encodeURIComponent(s);
      }

      this.excludeTrashedEmailsForDraftsAndSent = function (query) {
        if (query === 'tag:"drafts"' || query === 'tag:"sent"') {
          return query + ' -in:"trash"';
        }
        return query;
      };

      this.refreshMails = function () {
        var url = this.attr.mailsResource + '?q=' + escaped(this.attr.lastQuery) + '&p=' + this.attr.currentPage + '&w=' + this.attr.pageSize;

        this.attr.lastQuery = this.excludeTrashedEmailsForDraftsAndSent(this.attr.lastQuery);

        monitoredAjax(this, url, { dataType: 'json' })
          .done(function (data) {
            this.attr.numPages = Math.ceil(data.stats.total / this.attr.pageSize);
            this.trigger(document, events.mails.available, _.merge(_.merge({tag: this.attr.currentTag }), this.parseMails(data)));
          }.bind(this))
          .fail(function () {
            this.trigger(document, events.ui.userAlerts.displayMessage, { message: i18n('Could not fetch messages') });
          }.bind(this));
      };

      function createSingleMailUrl(mailsResource, ident) {
        return mailsResource + '/' + ident;
      }

      this.fetchSingle = function (event, data) {
        var fetchUrl = createSingleMailUrl(this.attr.singleMailResource, data.mail);

        monitoredAjax(this, fetchUrl, { dataType: 'json' })
          .done(function (mail) {
            if (_.isNull(mail)) {
              this.trigger(data.caller, events.mail.notFound);
              return;
            }

            this.trigger(data.caller, events.mail.here, { mail: this.mailFromJSON(mail) });
          }.bind(this));
      };

      this.previousPage = function () {
        if (this.attr.currentPage > 1) {
          this.updateCurrentPageNumber(this.attr.currentPage - 1);
          this.refreshMails();
        }
      };

      this.nextPage = function () {
        if (this.attr.currentPage < (this.attr.numPages)) {
          this.updateCurrentPageNumber(this.attr.currentPage + 1);
          this.refreshMails();
        }
      };

      this.updateCurrentPageNumber = function (newCurrentPage) {
        this.attr.currentPage = newCurrentPage;
        this.trigger(document, events.ui.page.changed, {
          currentPage: this.attr.currentPage,
          numPages: this.attr.numPages
        });
      };

      this.wantDraftReplyForMail = function (ev, data) {
        if (!features.isEnabled('draftReply')) {
          this.trigger(document, events.mail.draftReply.notFound);
          return;
        }

        monitoredAjax(this, '/draft_reply_for/' + data.ident, { dataType: 'json' })
          .done(function (mail) {
            if (_.isNull(mail)) {
              this.trigger(document, events.mail.draftReply.notFound);
              return;
            }
            this.trigger(document, events.mail.draftReply.here, { mail: this.mailFromJSON(mail) });
          }.bind(this));
      };

      this.after('initialize', function () {
        that = this;

        if (features.isEnabled('tags')) {
          this.on(events.mail.tags.update, this.updateTags);
        }

        this.on(document, events.mail.draftReply.want, this.wantDraftReplyForMail);
        this.on(document, events.mail.want, this.fetchSingle);
        this.on(document, events.mail.read, this.readMail);
        this.on(document, events.mail.unread, this.unreadMail);
        this.on(document, events.mail.delete, this.deleteMail);
        this.on(document, events.mail.deleteMany, this.deleteManyMails);
        this.on(document, events.search.perform, this.newSearch);
        this.on(document, events.ui.tag.selected, this.fetchByTag);
        this.on(document, events.ui.tag.select, this.fetchByTag);
        this.on(document, events.ui.mails.refresh, this.refreshMails);
        this.on(document, events.ui.page.previous, this.previousPage);
        this.on(document, events.ui.page.next, this.nextPage);

        this.fetchByTag(null, {tag: urlParams.getTag()});
      });
    }
  }
);
